### PR TITLE
Add Snyk report and Configraun

### PR DESCRIPTION
**Pull request by aug24**
## What does this change?

Adds a Snyk Report page
Fetches some configuration via Configraun configuration engine

## What is the value of this?

Exposes the results of snyk testing for vulnerabilities, including vulnerabilities found after the 
initial test.

Configraun use removes a 'secret' from the configuration file and stores it in AWS Parameter Store.
We should be moving towards this approach generally.  This implementation provides a dogfood implementation for others to use.

## Will this require CloudFormation and/or updates to the AWS StackSet?

Yes.
Template changes are included.

## Any additional notes?

No

### CODE REVIEW

**Comment by shtukas** [comment]
Looks good to me: +1

File: hq/test/logic/SnykDisplayTest.scala [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+
+  "fail to find organisationId (not nice)" in {
+    val organisationId = SnykDisplay.getOrganisationId(SnykDisplayTest.mockBadResponseWithoutMessage, Organisation("guardian"))
+    organisationId.isFailedAttempt shouldBe true
+    organisationId.getFailedAttempt.failures.head.friendlyMessage shouldBe """Could not read Snyk response ({"toughluck": "no use"})"""
+  }
+
+  "find project id list" in {
+    val projects = SnykDisplay.getProjectIdList(SnykDisplayTest.mockGoodProjectResponse)
+    !projects.value().exists(p => p.name == "project1") shouldBe false
```
Comment (adamnfish) [inline hq/test/logic/SnykDisplayTest.scala]: ScalaTest has nice ways to test properties of collections. As well as being terser and easier to read, you'll get better error messages when it fails. In this case it's not so simple because you're looking for a nested property, but I'll put this link here for your perusal!

http://www.scalatest.org/user_guide/using_matchers

File: hq/app/logic/SnykDisplay.scala [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    val projectIds = (Json.parse(s) \ "projects").asOpt[List[SnykProject]]
+    val error = Json.parse(s).asOpt[SnykError].getOrElse(SnykError(s))
+    val f = Failure(s"Unable to find project ids from $s", s"Could not read Snyk response (${error.error})", 502, None, None)
+    Attempt.fromOption(projectIds, FailedAttempt(f))
+  }
+
+  def parseProjectVulnerabilities(projects: List[String])(implicit ec:ExecutionContext): Attempt[List[SnykProjectIssues]] = {
+    val b = projects.map( s => {
+      val projectVulnerabilities = Json.parse(s).asOpt[SnykProjectIssues]
+      val error = Json.parse(s).asOpt[SnykError].getOrElse(SnykError(s))
```
[the remark this answers is not in the document]
Comment (adamnfish) [reply hq/app/logic/SnykDisplay.scala]: (Same for all of them)

File: hq/app/logic/SnykDisplay.scala [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+  def getProjectIdList(s: String): Attempt[List[SnykProject]] = {
+    val projectIds = (Json.parse(s) \ "projects").asOpt[List[SnykProject]]
+    val error = Json.parse(s).asOpt[SnykError].getOrElse(SnykError(s))
+    val f = Failure(s"Unable to find project ids from $s", s"Could not read Snyk response (${error.error})", 502, None, None)
+    Attempt.fromOption(projectIds, FailedAttempt(f))
+  }
+
+  def parseProjectVulnerabilities(projects: List[String])(implicit ec:ExecutionContext): Attempt[List[SnykProjectIssues]] = {
+    val b = projects.map( s => {
```
[the remark this answers is not in the document]
Comment (adamnfish) [reply hq/app/logic/SnykDisplay.scala]: Same with all the `b`s and `a`s around here. `a` and `b` are respectively good names for values with type `A` and `B`, but they are not so helpful when referring to something concrete!

File: hq/test/logic/SnykDisplayTest.scala [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+    projects.getFailedAttempt.failures.head.friendlyMessage shouldBe "Could not read Snyk response (some nice error)"
+  }
+
+  "fail to find vulnerability list (not nice)" in {
+    val projects = SnykDisplay.parseProjectVulnerabilities(List(SnykDisplayTest.mockBadResponseWithoutMessage))
+    projects.isFailedAttempt() shouldBe true
+    projects.getFailedAttempt.failures.head.friendlyMessage shouldBe """Could not read Snyk response ({"toughluck": "no use"})"""
+  }
+
+  "label projects" in {
```
Comment (adamnfish) [inline hq/test/logic/SnykDisplayTest.scala]: This test message isn't very useful! What does it do / what property are we checking?

File: hq/markdown/temporary-ssh.md [hunk 5]
```diff
@@ -0,0 +1,29 @@
+# Temporary SSH
+
+In order to move towards use of temporary credentials, a facility has been provided to locally generate a keypair and push the
```
Comment (adamnfish) [inline hq/markdown/temporary-ssh.md]: This is a bit waffly!

File: hq/markdown/temporary-ssh.md [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+## Implementation
+
+The mechanism is via AWS run-command.  Please see AWS documentation for details.
+
+## Pre-requisites
+
+The instance must 
+
+  * be set up to use run-command
```
Comment (adamnfish) [inline hq/markdown/temporary-ssh.md]: People will ask how they do this, I think

File: hq/markdown/parameter-store.md [hunk 7]
```diff
@@ -0,0 +1,57 @@
+# SSM Parameters
```
Comment (adamnfish) [inline hq/markdown/parameter-store.md]: If this is documenting the Guardian wrapper library, can we just point people at its README?

File: hq/app/views/snyk/snyk.scala.html [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+            </thead>
+
+            <tbody>
+            @for(project <- results) {
+                <tr>
+                    <td>
+                        <a class="black-text" href="https://snyk.io/org/guardian/project/@project.id/">@project.name</a>
+                    </td>
+                    <td>
+                        <a class="black-text" href="https://snyk.io/org/guardian/project/@project.id/?disclosure=all&severity=high">
```
Comment (adamnfish) [inline hq/app/views/snyk/snyk.scala.html]: If the org is coming from config, should that be reflected here?

File: hq/app/model/models.scala [hunk 9]
```diff
@@ -135,3 +135,21 @@ case class MachineUser  (
   reportStatus: ReportStatus,
   lastActivityDay : Option[Long]
 )
+
+case class Token(value: String) extends AnyVal
```
Comment (adamnfish) [inline hq/app/model/models.scala]: Token is quite a general name. Is this really `SnykToken`?

File: build.sbt [hunk 10]
```diff
@@ -10,6 +10,7 @@ scalaVersion in ThisBuild := "2.12.3"
 scalacOptions in ThisBuild ++= Seq("-deprecation", "-feature", "-unchecked", "-target:jvm-1.8", "-Xfatal-warnings")
 
 // resolvers += "guardian-bintray" at "https://dl.bintray.com/guardian/sbt-plugins/"
+resolvers += DefaultMavenRepository
```
Comment (adamnfish) [inline build.sbt]: Why is this required?

File: hq/app/logic/SnykDisplay.scala [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+      val projectVulnerabilities = Json.parse(s).asOpt[SnykProjectIssues]
+      val error = Json.parse(s).asOpt[SnykError].getOrElse(SnykError(s))
+      val f = Failure(s"Unable to find project vulnerabilities from $s", s"Could not read Snyk response (${error.error})", 502, None, None)
+      Attempt.fromOption(projectVulnerabilities, FailedAttempt(f))
+    })
+    Attempt.traverse(b)(a => a)
+  }
+
+  def labelProjects(projects: List[SnykProject], responses: List[SnykProjectIssues]): List[SnykProjectIssues] = {
+    projects.zip(responses).map(a => a._2.copy(name = a._1.name).copy(id = a._1.id))
```
[the remark this answers is not in the document]
Comment (adamnfish) [reply hq/app/logic/SnykDisplay.scala]: (it might also let you use a helpful name)

File: build.sbt [hunk 12]
```diff
@@ -22,10 +23,11 @@ lazy val hq = (project in file("hq")).
     libraryDependencies ++= Seq(
       ws,
       filters,
-      "com.gu" %% "play-googleauth" % "0.7.0",
+      "com.gu" %% "play-googleauth" % "0.7.1",
       "joda-time" % "joda-time" % "2.9.9",
-      "org.typelevel" %% "cats" % "0.8.1",
+      "org.typelevel" %% "cats-core" % "1.0.1",
```
[the remark this answers is not in the document]
Comment (aug24) [reply build.sbt]: I liaised with the chaps upversioning play-googleauth.  I believe it's correct.

File: hq/app/api/Snyk.scala [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+      .get
+
+    Attempt.fromFuture(futureResponse) { case NonFatal(e) =>
+      val failure = Failure(e.getMessage, "Could not read organisations from Snyk", 502, None, Some(e))
+      FailedAttempt(failure)
+    }
+  }
+
+  def getProjects(token: Token, id: String, wsClient: WSClient)(implicit ec:ExecutionContext): Attempt[WSResponse] = {
+        val snykProjectsUrl = s"https://snyk.io/api/v1/org/$id/projects"
```
[the remark this answers is not in the document]
Comment (aug24) [reply hq/app/api/Snyk.scala]: Done.

File: hq/markdown/temporary-ssh.md (hunk 6 above)
Comment (katebee) [reply hq/markdown/temporary-ssh.md]: could we be specific: `EC2 run command` or even `AWS EC2 run command`

It may help if we wrap this in a link to the SHQ documentation on EC2 run (will relative links work here?)

File: hq/app/views/snyk/snyk.scala.html [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+                    <th>Medium</th>
+                    <th>Low</th>
+                </tr>
+            </thead>
+
+            <tbody>
+            @for(project <- results) {
+                <tr>
+                    <td>
+                        <a class="black-text" href="https://snyk.io/org/guardian/project/@project.id/">@project.name</a>
```
[the remark this answers is not in the document]
Comment (katebee) [reply hq/app/views/snyk/snyk.scala.html]: I have also not loaded up the UI (yet). However black links will be bad for a11y and discoverability.

I would make an exception if this is a card (as with the index page) where the entire card is clickable and therefore we may want black text on the card (see usage of `hp__link`)

If the layout is going to be similar to the IAM page, maybe that design pattern could be used?

picture 168

[the remark this answers is not in the document]
Comment (katebee) [reply hq/app/views/snyk/snyk.scala.html]: Opinionated (IANAUX/D):

Black text on cards and within nav is okay (and is the current pattern), since most users will expect those areas to be clickable. However, within a table or body text, links should be visually distinct (at least a different colour, ideally text decoration).

Regarding accessibility: 

> To aid discoverability all links must be made self-evident with their visual style

http://www.bbc.co.uk/guidelines/futuremedia/accessibility/html/control-styles.shtml

File: hq/app/AppComponents.scala [hunk 15]
```diff
@@ -24,13 +28,37 @@ class AppComponents(context: Context)
...
+          case Left(a) =>
+            Logger.error(s"Unable to load Configraun configuration from AWS (${a.message})")
+            sys.exit(1)
+          case Right(a: com.gu.configraun.models.Configuration) => a
+        }
+      case _ => Configraun.loadConfig match {
+        case Left(a) =>
+          Logger.error(s"Unable to load Configraun configuration from AWS tags (${a.message})")
+          sys.exit(1)
+        case Right(a: com.gu.configraun.models.Configuration) => a
```
[the remark this answers is not in the document]
Comment (aug24) [reply hq/app/AppComponents.scala]: No.  Done.

File: hq/test/logic/SnykDisplayTest.scala [hunk 16]
```diff
@@ -0,0 +1,211 @@
+package logic
+
+import model._
+import org.scalatest.{FreeSpec, Matchers}
+import utils.attempt.AttemptValues
+import scala.concurrent.ExecutionContext.Implicits.global
+
+class SnykDisplayTest extends FreeSpec with Matchers with AttemptValues {
```
[the remark this answers is not in the document]
Comment (aug24) [reply hq/test/logic/SnykDisplayTest.scala]: Done.

File: hq/app/AppComponents.scala [hunk 17]
```diff
@@ -24,13 +28,37 @@ class AppComponents(context: Context)
...
+
+    configuration.getOptional[String]("stage") match {
+      case Some("DEV") =>
+        val stack = configuration.get[String]("stack")
+        val app = configuration.get[String]("app")
+        val stage = "DEV"
+        Configraun.loadConfig(Identifier(Stack(stack), App(app), Stage.fromString(stage).get)) match {
+          case Left(a) =>
+            Logger.error(s"Unable to load Configraun configuration from AWS (${a.message})")
+            sys.exit(1)
```
Comment (adamnfish) [inline hq/app/AppComponents.scala]: O_O

Why are we calling sys.exit in a webserver?!

Comment (aug24) [reply hq/app/AppComponents.scala]: Fast fail in the event of bad config. Perhaps a comment would be useful, or there may be a better (more Play) answer, but the action is correct IMO.
